File: src/tax/us/form_8949.py
"""
IRS Form 8949 generator - standalone functions for the tax module system.

Adapted from src/reports/form_8949.py to work with Disposal objects directly
(instead of DB rows), enabling use from the TaxModule.generate_reports() flow.

Box codes (per IRS instructions):
  A = short-term, reported on 1099-B with basis
  B = short-term, reported on 1099-B without basis
  C = short-term, NOT reported on 1099-B (self-reported)
  D = long-term, reported on 1099-B with basis
  E = long-term, reported on 1099-B without basis
  F = long-term, NOT reported on 1099-B (self-reported)
"""
from __future__ import annotations

import csv
import io
from datetime import datetime
from decimal import Decimal
from typing import Optional

from src.tax.models import HoldingPeriod

# Brokers known to issue 1099-DA
_1099_DA_BROKERS: frozenset[str] = frozenset({
    "coinbase", "kraken", "gemini", "binance.us", "robinhood", "paypal",
    "cash app", "etoro",
})


def _fmt_date(dt) -> str:
    """Convert datetime or ISO string to MM/DD/YYYY."""
    if isinstance(dt, datetime):
        return dt.strftime("%m/%d/%Y")
    if isinstance(dt, str):
        try:
            return datetime.fromisoformat(dt[:10]).strftime("%m/%d/%Y")
        except (ValueError, TypeError):
            return dt
    return str(dt)


def _box_code(holding_period: str, is_reported: bool, has_basis: bool) -> str:
    """Determine the Form 8949 checkbox code."""
    is_short = holding_period in ("short-term", "short", "SHORT_TERM")
    if is_short:
        if is_reported and has_basis:
            return "A"
        if is_reported:
            return "B"
        return "C"
    # long-term or exempt (exempt disposals in US shouldn't happen, but handle gracefully)
    if is_reported and has_basis:
        return "D"
    if is_reported:
        return "E"
    return "F"


def generate_form_8949_csv(
    disposals: list,
    year: int,
    method: str,
    broker_map: Optional[dict[str, str]] = None,
) -> str:
    """
    Generate Form 8949 CSV from Disposal objects.

    Args:
        disposals: list of Disposal dataclass instances
        year: tax year to filter
        method: cost basis method label
        broker_map: tx_hash -> broker_name for 1099-DA flagging

    Returns:
        CSV string
    """
    broker_map = broker_map or {}

    output = io.StringIO()
    writer = csv.writer(output)

    headers = [
        "Part", "Box", "Description", "Date Acquired", "Date Sold",
        "Proceeds", "Cost Basis", "Adjustment Code", "Adjustment Amount",
        "Gain or Loss", "1099-DA Reported",
    ]
    writer.writerow(headers)

    for d in disposals:
        # Filter by year
        disposal_date = d.date if hasattr(d, "date") else d.get("date")
        if hasattr(disposal_date, "year"):
            if disposal_date.year != year:
                continue
        else:
            continue

        broker = broker_map.get(d.tx_hash, "").lower()
        is_reported = broker in _1099_DA_BROKERS
        has_basis = is_reported

        hp = d.holding_period
        if isinstance(hp, HoldingPeriod):
            hp_str = hp.value
        else:
            hp_str = str(hp)

        is_short = hp_str in ("short-term", "short", "SHORT_TERM")
        part = "I (Short-Term)" if is_short else "II (Long-Term)"
        box = _box_code(hp_str, is_reported, has_basis)

        description = f"{d.amount:.8f} {d.token}".rstrip("0").rstrip(".")

        writer.writerow([
            part, box, description,
            "VARIOUS", _fmt_date(d.date),
            str(d.proceeds_usd), str(d.cost_basis_usd),
            "", "0",
            str(d.gain_loss_usd),
            "Yes" if is_reported else "No",
        ])

    return output.getvalue()

File: frontend/src/engine/tax/us/form8949.ts
/**
 * IRS Form 8949 generator.
 *
 * Ported from src/tax/us/form_8949.py.
 *
 * Box codes (per IRS instructions):
 *   A = short-term, reported on 1099-B with basis
 *   B = short-term, reported on 1099-B without basis
 *   C = short-term, NOT reported on 1099-B (self-reported)
 *   D = long-term, reported on 1099-B with basis
 *   E = long-term, reported on 1099-B without basis
 *   F = long-term, NOT reported on 1099-B (self-reported)
 *
 * CRITICAL: All monetary values use decimal.js Decimal -- NEVER native number for money.
 */

import Decimal from 'decimal.js';
import type { Disposal, ReportFile } from '../../types';
import { buildCsv } from '../../reports/csv-export';

// Brokers known to issue 1099-DA
const _1099_DA_BROKERS = new Set([
  'coinbase', 'kraken', 'gemini', 'binance.us', 'robinhood', 'paypal',
  'cash app', 'etoro',
]);

/** Convert ISO date string to MM/DD/YYYY. */
function fmtDateUS(dt: string): string {
  try {
    const d = new Date(dt.slice(0, 10));
    if (isNaN(d.getTime())) return dt;
    const mm = String(d.getUTCMonth() + 1).padStart(2, '0');
    const dd = String(d.getUTCDate()).padStart(2, '0');
    const yyyy = d.getUTCFullYear();
    return `${mm}/${dd}/${yyyy}`;
  } catch {
    return dt;
  }
}

function isShortTerm(holdingPeriod: string): boolean {
  return ['short-term', 'short', 'SHORT_TERM'].includes(holdingPeriod);
}

/** Determine the Form 8949 checkbox code. */
function boxCode(holdingPeriod: string, isReported: boolean, hasBasis: boolean): string {
  const short = isShortTerm(holdingPeriod);
  if (short) {
    if (isReported && hasBasis) return 'A';
    if (isReported) return 'B';
    return 'C';
  }
  if (isReported && hasBasis) return 'D';
  if (isReported) return 'E';
  return 'F';
}

/**
 * Generate IRS Form 8949 CSV from Disposal objects.
 *
 * @param disposals  - Array of Disposal objects
 * @param year       - Tax year to filter
 * @param brokerMap  - Optional txHash -> broker name for 1099-DA flagging
 * @returns          ReportFile with CSV content
 */
export function generateForm8949(
  disposals: Disposal[],
  year: number,
  brokerMap: Record<string, string> = {},
): ReportFile {
  const headers = [
    'Part', 'Box', 'Description', 'Date Acquired', 'Date Sold',
    'Proceeds', 'Cost Basis', 'Adjustment Code', 'Adjustment Amount',
    'Gain or Loss', '1099-DA Reported',
  ];

  const rows: string[][] = [];

  for (const d of disposals) {
    const disposalYear = parseInt(d.date.slice(0, 4), 10);
    if (disposalYear !== year) continue;

    const broker = (brokerMap[d.txHash] ?? '').toLowerCase();
    const isReported = _1099_DA_BROKERS.has(broker);
    const hasBasis = isReported;

    const hp = d.holdingPeriod;
    const short = isShortTerm(hp);
    const part = short ? 'I (Short-Term)' : 'II (Long-Term)';
    const box = boxCode(hp, isReported, hasBasis);

    // Format amount: strip trailing zeros
    const description = `${d.amount.toFixed(8).replace(/0+$/, '').replace(/\.$/, '')} ${d.token}`;

    rows.push([
      part,
      box,
      description,
      'VARIOUS',
      fmtDateUS(d.date),
      d.proceedsUsd.toString(),
      d.costBasisUsd.toString(),
      '',
      '0',
      d.gainLossUsd.toString(),
      isReported ? 'Yes' : 'No',
    ]);
  }

  return {
    filename: `form_8949_${year}.csv`,
    content: buildCsv(headers, rows),
    mimeType: 'text/csv',
    reportType: 'form_8949',
  };
}
